internal/workarounds/prom.go

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package workarounds
+
+import (
+	"context"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8swait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	promConfigMapNamespace = "openshift-monitoring"
+	promConfigMapName      = "cluster-monitoring-config"
+	promConfigMapKey       = "config.yaml"
+
+	promConfigKey       = "prometheusK8s"
+	promConfigVolumeKey = "volumeClaimTemplate"
+)
+
+type Prometheus struct {
+	PollInterval time.Duration
+	PollTimeout  time.Duration
+}
+
+func ForPrometheus() *Prometheus {
+	return &Prometheus{
+		PollInterval: 5 * time.Second,
+		PollTimeout:  5 * time.Minute,
+	}
+}
+
+func (prom *Prometheus) IssueReference() string {
+	return "https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/1784"
+}
+
+func (prom *Prometheus) Describe() string {
+	return "Prometheus instance  cannot use storage and fails to go running; only one instance out of two will go running, and this later prevents node draining"
+}
+
+func (prom *Prometheus) Apply(ctx context.Context, cli client.Client) error {
+	var cm corev1.ConfigMap
+	key := client.ObjectKey{
+		Namespace: promConfigMapNamespace,
+		Name:      promConfigMapName,
+	}
+	immediate := true
+	return k8swait.PollUntilContextTimeout(ctx, prom.PollInterval, prom.PollTimeout, immediate, func(ctx2 context.Context) (bool, error) {
+		err := cli.Get(ctx2, key, &cm)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				klog.InfoS("configmap not found, nothing to do", "namespace", key.Namespace, "name", key.Name)
+				return true, nil
+			}
+			return true, err
```

Comment:
do you want the loop to be aborted if you find another error? if not you need to return nil instead
and regardless why to return true? the condition isn't satisfied

Reply:
yes, this failure is critical and needs to make the process abort